chainindex/indexer.go

```diff
@@ -0,0 +1,475 @@
+	// Height(to) < Height(from)
+
+	revertTsKeyCid, err := toTipsetKeyCidBytes(from)
+	if err != nil {
+		return xerrors.Errorf("error getting tipset key cid: %w", err)
+	}
+
+	err = withTx(ctx, si.db, func(tx *sql.Tx) error {
+		if _, err := tx.Stmt(si.revertTipsetStmt).ExecContext(ctx, revertTsKeyCid); err != nil {
+			return xerrors.Errorf("error marking tipset as reverted: %w", err)
+		}
+
+		// index the `to` tipset as it has now been applied -> simply for redundancy
+		if err := si.indexTipset(ctx, tx, to); err != nil {
+			return xerrors.Errorf("error indexing tipset: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return xerrors.Errorf("error during revert transaction: %w", err)
+	}
+
+	si.notifyUpdateSubs()
+
+	return nil
+}
+
+func (si *SqliteIndexer) indexTipset(ctx context.Context, tx *sql.Tx, ts *types.TipSet) error {
+	tsKeyCidBytes, err := toTipsetKeyCidBytes(ts)
+	if err != nil {
+		return xerrors.Errorf("error computing tipset cid: %w", err)
+	}
+
+	restored, err := si.restoreTipsetIfExists(ctx, tx, tsKeyCidBytes)
+	if err != nil {
+		return xerrors.Errorf("error restoring tipset: %w", err)
+	}
+	if restored {
+		return nil
+	}
+
+	height := ts.Height()
+	insertTipsetMsgStmt := tx.Stmt(si.insertTipsetMessageStmt)
+
+	msgs, err := si.cs.MessagesForTipset(ctx, ts)
+	if err != nil {
+		return xerrors.Errorf("error getting messages for tipset: %w", err)
+	}
+
+	if len(msgs) == 0 {
+		// If there are no messages, just insert the tipset and return
+		if _, err := insertTipsetMsgStmt.ExecContext(ctx, tsKeyCidBytes, height, 0, nil, -1); err != nil {
+			return xerrors.Errorf("error inserting empty tipset: %w", err)
+		}
+		return nil
+	}
+
+	for i, msg := range msgs {
+		msg := msg
+		if _, err := insertTipsetMsgStmt.ExecContext(ctx, tsKeyCidBytes, height, 0, msg.Cid().Bytes(), i); err != nil {
+			return xerrors.Errorf("error inserting tipset message: %w", err)
+		}
+	}
+
+	for _, blk := range ts.Blocks() {
+		blk := blk
+		_, smsgs, err := si.cs.MessagesForBlock(ctx, blk)
+		if err != nil {
+			return err
+		}
+
+		for _, smsg := range smsgs {
+			smsg := smsg
+			if smsg.Signature.Type != crypto.SigTypeDelegated {
+				continue
+			}
+			if err := si.indexSignedMessage(ctx, tx, smsg); err != nil {
+				return xerrors.Errorf("error indexing eth tx hash: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) restoreTipsetIfExists(ctx context.Context, tx *sql.Tx, tsKeyCidBytes []byte) (bool, error) {
+	// Check if the tipset already exists
+	var exists bool
+	if err := tx.Stmt(si.tipsetExistsStmt).QueryRowContext(ctx, tsKeyCidBytes).Scan(&exists); err != nil {
+		return false, xerrors.Errorf("error checking if tipset exists: %w", err)
+	}
+	if exists {
+		if _, err := tx.Stmt(si.tipsetUnRevertStmt).ExecContext(ctx, tsKeyCidBytes); err != nil {
+			return false, xerrors.Errorf("error restoring tipset: %w", err)
+		}
+		return true, nil
+	}
```

Comment:
I _think_ you could do this in a single query, something like:

```sql
UPDATE tipset_message
  SET reverted =
    CASE WHEN reverted != 0 THEN 0 ELSE reverted END
  WHERE tipset_key_cid = ? AND reverted != 0
```

and then check `RowsAffected` is `>0`.

Reply:
Want to keep it simple. Anyways, this should only be used when a given tipset is applied -> reverted -> applied again i.e. not often.